### PR TITLE
feat: postgres - ch sync to support groups

### PR DESCRIPTION
**Pull request by tiina303**
## Problem

Groups can also be out of sync, longer term we should look into adding version there, but for now we are using `argMax(group_properties, _timestamp) ` in queries.

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: posthog/management/commands/test/test_sync_persons_to_clickhouse.py [hunk 1]
```diff
@@ -106,6 +114,52 @@ def test_distinct_ids_deleted(self):
...
+            group_properties={"a": 1234},
+            created_at=ts,
+            version=5,
+        )
+
+        run_group_sync(self.team.pk, live_run=True, sync=True)
+
+        ch_groups = sync_execute(
+            """
+            SELECT group_type_index, group_key, group_properties, created_at FROM groups WHERE team_id = %(team_id)s
```
Comment (hazzadous) [inline posthog/management/commands/test/test_sync_persons_to_clickhouse.py:132]: Can we check that version is also set appropriately as this is likely key to the ClickHouse merge functionality?

File: posthog/management/commands/test/test_sync_persons_to_clickhouse.py [hunk 2]
```diff
@@ -106,6 +114,52 @@ def test_distinct_ids_deleted(self):
         )
         self.assertEqual(ch_person_distinct_ids, [(UUID(int=0), self.team.pk, "test-id-7", 107, True)])
 
+    def test_group_sync(self):
```
Comment (hazzadous) [inline posthog/management/commands/test/test_sync_persons_to_clickhouse.py:117]: Can we validate that running `run_group_sync` twice doesn't result in another update?

File: posthog/management/commands/test/test_sync_persons_to_clickhouse.py [hunk 3]
```diff
@@ -106,6 +114,52 @@ def test_distinct_ids_deleted(self):
...
+            {"team_id": self.team.pk},
+        )
+        self.assertEqual(len(ch_groups), 1)
+        ch_group = ch_groups[0]
+        self.assertEqual(ch_group[0], 2)
+        self.assertEqual(ch_group[1], "group-key")
+        self.assertEqual(ch_group[2], '{"a": 1234}')
+        self.assertEqual(ch_group[3].strftime("%Y-%m-%d %H:%M:%S"), ts.strftime("%Y-%m-%d %H:%M:%S"))
+
+    def test_group_sync_updates_group(self):
```
Comment (hazzadous) [inline posthog/management/commands/test/test_sync_persons_to_clickhouse.py:143]: Can we validate that running run_group_sync twice doesn't result in another update?

File: posthog/management/commands/sync_persons_to_clickhouse.py [hunk 4]
```diff
@@ -139,9 +145,47 @@ def run_distinct_id_sync(team_id: int, live_run: bool, deletes: bool, sync: bool
...
             if distinct_id not in postgres_distinct_ids:
                 logger.info(f"Deleting distinct ID {distinct_id}")
                 if live_run:
                     _delete_ch_distinct_id(team_id, UUID(int=0), distinct_id, version, sync=sync)
+
+
+def run_group_sync(team_id: int, live_run: bool, sync: bool):
+    logger.info("Running group table sync")
+    # lookup what needs to be updated in ClickHouse and send kafka messages for only those
+    pg_groups = Group.objects.filter(team_id=team_id)
```
Comment (hazzadous) [inline posthog/management/commands/sync_persons_to_clickhouse.py:160]: Do we need all the fields here? Could we limit the number of fields?

File: posthog/management/commands/sync_persons_to_clickhouse.py [hunk 5]
```diff
@@ -139,9 +145,47 @@ def run_distinct_id_sync(team_id: int, live_run: bool, deletes: bool, sync: bool
...
+                    group_type_index=pg_group.group_type_index,
+                    group_key=pg_group.group_key,
+                    properties=pg_group.group_properties,
+                    created_at=pg_group.created_at,
+                    sync=sync,
+                )
+
+
+def should_update_group(ch_group, pg_group: Group) -> bool:
+    return pg_group.group_properties != ch_group["properties"] or pg_group.created_at != ch_group["created_at"]
```
Comment (hazzadous) [inline posthog/management/commands/sync_persons_to_clickhouse.py:191]: Given that we have `version`, do we need to do a deep comparison here?

File: posthog/management/commands/test/test_sync_persons_to_clickhouse.py (hunk 1 above)
Comment (tiina303) [reply posthog/management/commands/test/test_sync_persons_to_clickhouse.py:132]: there is no version, we're doing this based on `_timestamp`, can check that

File: posthog/management/commands/test/test_sync_persons_to_clickhouse.py (hunk 3 above)
Comment (tiina303) [reply posthog/management/commands/test/test_sync_persons_to_clickhouse.py]: good call cause it totally did 🙈

**Review by hazzadous — APPROVED** [review]
Approved although I think we should look at better ways to keep these in sync.

It looks like there are race conditions still in the current code, i.e. if a group is updated after this script has read the group, but we generate the _timestamp as a value after whatever that other process used, then we'll end up with ClickHouse choosing the wrong value.

As an immediate improvement if we could get PostgreSQL to give us a timestamp of the latest update to the group that would at least ensure that we will order versions in ClickHouse accordingly. We still have chance of timestamp collisions but it would be better than nothing.

I more robust solution may be to, rather than send the group data from multiple places which might get out of sync, instead:

1. when we update a group, simply send the group key to the topic, key'd on group key such that all messages for a group are in the same partition.
2. have a consumer reading from this topic (perhaps deduplicating 1 minute batches of update requests so you always know that there would only be max one read from PostgreSQL per group key), reading from PostgreSQL the data and inserting into ClickHouse.

This coupled with ensuring setting offsets _after_ we've persisted the group update notification to Kafka would ensure the two systems don't go out of sync.

**Comment by tiina303** [comment]
Yup agreed on the race conditions here - I think adding version to the groups table in CH would be the way to go, it already exists on Postgres side

**Pull request merged**